Include Flink-Job-Cluster Helm Chart

yalctay93 opened the pull request:
Added a helm-chart for flink-job-cluster deployment.

The chart was tested solely for our own use cases, which might differ from yours.

<sub> Tayfun Yalcin <tayfun.yalcin@daimler.com>, Daimler TSS GmbH, [Legal footer / Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)</sub>

yalctay93 commented:
we signed a cla with the following google group "dai-google-cla"
my email address that is part of the commits is added there as a member and I have added it to my github profile

I wonder why the cla/google check has failed

functicons commented:
@yalctay93, thanks for the PR! Could you let me know the purpose and use case of this helm chart?

yalctay93 commented:
> @yalctay93, thanks for the PR! Could you let me know the purpose and use case of this helm chart?

Hello! We always deploy ressources with helm3 in our project, it is part of our CI/CD flow. That is why we created a helm-chart for the flink-job-cluster. Obviously we could've used the flink-job-cluster without helm charts but it gives us more advantages such as automated rollbacks, deployment checks, dry runs etc. Whenever one of our devs has to spin up a new job cluster, they'll only configure the values.yaml and deploy the chart which is also more convenient for them.

functicons commented:
Sounds great. It would be nice if you can add a doc for the usage in this [dir](https://github.com/GoogleCloudPlatform/flink-on-k8s-operator/tree/master/docs), add add a link to it from [here](https://github.com/GoogleCloudPlatform/flink-on-k8s-operator#installation).

functicons commented:
If you see fit, could you add your company to this list? Thanks! https://github.com/GoogleCloudPlatform/flink-on-k8s-operator/blob/master/docs/who_is_using.md

yalctay93 commented:
> Sounds great. It would be nice if you can add a doc for the usage in this [dir](https://github.com/GoogleCloudPlatform/flink-on-k8s-operator/tree/master/docs), and add a link to it from [here](https://github.com/GoogleCloudPlatform/flink-on-k8s-operator#installation).

Sure, I'll add it.

yalctay93 commented:
I've updated the docs and added the guide for flink-job-cluster helm chart. I hope it is enough, if something is unclear please let me know so I can adjust it.

functicons commented:
Thanks for the update! I'll test it tomorrow before giving you feedback.

functicons commented on README.md:
```diff
@@ -81,6 +81,7 @@ The operator is still under active development, there is no Helm chart available
 * [Run Apache Beam Python jobs](docs/beam_guide.md)
 * [Use GCS connector](images/flink/README.md)
 * [Test with Apache Kafka](docs/kafka_test_guide.md)
+* [Use flink-job-cluster helm-chart](docs/flink_job_cluster_guide.md)
```
Create Flink job clusters with Helm Chart

functicons commented on docs/flink_job_cluster_guide.md:
```diff
@@ -0,0 +1,95 @@
+# Running Flink-Job-Cluster in Kubernetes
```
Create Flink job clusters with Helm Chart

functicons commented on docs/flink_job_cluster_guide.md:
```diff
@@ -0,0 +1,95 @@
+
+## Overview
+
+to run a Flink-Job-Cluster, you first have to deploy the Flink-Operator. For that, follow the [user_guide](https://github.com/GoogleCloudPlatform/flink-on-k8s-operator/blob/master/docs/user_guide.md)
```
Mention that this Helm Chart is an addition to the existing way of creating Flink job clusters. Don't need to mention deploying the operator since you will mention it in the Prerequisites section.

functicons commented on docs/flink_job_cluster_guide.md:
```diff
@@ -0,0 +1,95 @@
+
+to run a Flink-Job-Cluster, you first have to deploy the Flink-Operator. For that, follow the [user_guide](https://github.com/GoogleCloudPlatform/flink-on-k8s-operator/blob/master/docs/user_guide.md)
+
+## Prerequisites
```
Make prerequisite as a list.
- Operator image version, deploying the operator 
- Flink image version.
- Helm version.

functicons commented on docs/flink_job_cluster_guide.md:
````diff
@@ -0,0 +1,95 @@
+
+The instructions to install the Flink Job Cluster chart:
+
+1. Prepare a Flink operator image. You should use at least ```gcr.io/flink-operator/flink-operator:v1beta1-6```.
````
The first 3 items should be part of the prerequisites.

functicons commented on docs/flink_job_cluster_guide.md:
````diff
@@ -0,0 +1,95 @@
+
+3. Prepare a Flink-Job-Cluster image. You should use ```flink:latest``` or ```flink:1.9.3``` which has been tested and confirmed to work with flink-job-cluster helm chart.
+
+3. Use the following command to install the flink-job-cluster chart:
````
From which directory to run the command?

functicons commented on docs/flink_job_cluster_guide.md:
````diff
@@ -0,0 +1,95 @@
+3. Use the following command to install the flink-job-cluster chart:
+
+  ```bash
+  helm template --namespace [NAMESPACE] [RELEASE_NAME] [CHART] -f values.yaml |kubectl apply -f -
````
IIUC, this `values.yaml` file is an example, so it it not part of the helm chart, it should be under the [examples/](https://github.com/GoogleCloudPlatform/flink-on-k8s-operator/tree/master/examples) dir.

Mention that this is an example, and elaborate what needs to be included in the `values.yaml` file, i.e., the spec.

yalctay93 replied:
Thank you for reviewing the guide! I have adjusted almost all of your suggestions however I am not entirely sure whether or not the values.yaml is just an example. It should be included in the helm-chart itself because it has the minimum configuration settings that are needed to succesfully deploy the Flink job cluster in a reliable state. As far as I know, values.yaml is always included in the helm-chart folders. Please correct me if I am wrong. We can also put an example at examples/dir but in my opinion it would be better to just include it in the chart.

functicons commented on docs/flink_job_cluster_guide.md:
````diff
@@ -0,0 +1,95 @@
+  helm template --namespace [NAMESPACE] [RELEASE_NAME] [CHART] -f values.yaml |kubectl apply -f -
+  ```
+
+## Uninstalling the Chart
````
Add a section for checking the release status.

yalctay93 replied:
Hello! I was not entirely sure what you meant with "release status" so I added a section to check the current Flink Operator Image Releases. If you meant the helm-chart release status, then that wouldn't be possible as the guide is meant to deploy the Flink job cluster with the helm template command and use a piped kubectl apply at the end. This way, we only use the templating of helm but not its releases.

functicons replied:
Gotcha. I meant to say Helm Chart release status, now I see it is not deployed directly.

functicons commented on docs/flink_job_cluster_guide.md:
```diff
@@ -0,0 +1,95 @@
+
+## Monitoring
+
+The Flink-Job-Cluster comes with a PodMonitor ressource, which is the counter part to a ServiceMonitor.
```
s/Flink-Job-Cluster/Flink job cluster, and all the occurrences.

functicons replied:
s/ressource/resource

functicons commented on docs/flink_job_cluster_guide.md:
````diff
@@ -0,0 +1,95 @@
+This way, you can directly download your jar files from a blob storage at Microsoft Azure.
+A basic script that is confirmed to work could look like this:
+
+```bash
````
s/bash/python

functicons commented on docs/flink_job_cluster_guide.md:
```diff
@@ -0,0 +1,95 @@
+
+You can use the following [dashboard](https://grafana.com/grafana/dashboards/10369) in your grafana to monitor the flink-job-cluster.
+
+## Run Jobs with InitContainer
```
This doesn't seem to be tightly relevant to this helm chart. You can briefly mention it and point to an example like https://github.com/GoogleCloudPlatform/flink-on-k8s-operator/blob/master/config/samples/flinkoperator_v1beta1_remotejobjar.yaml.

functicons commented on docs/flink_job_cluster_guide.md:
```diff
@@ -0,0 +1,95 @@
+
+## Updating Job
+
+Since the flink-job-cluster is only deployed while the job is running, you have to re-deploy the helm chart with the new JAR file ready to be downloaded and started.
```
This section is a little confusing to me. Could you rewrite is in a way that imagine a user has a running job and they want to update it, what should they do? Command line example would be helpful.

functicons commented on docs/flink_job_cluster_guide.md:
```diff
@@ -0,0 +1,108 @@
+
+## Overview
+
+This Helm Chart is an addition to the [existing](https://github.com/GoogleCloudPlatform/flink-on-k8s-operator/tree/master/config/samples) way of deploying Flink job clusters.
```
Mention the benefit provided by this Helm Chart over creating a YAML directly, so that users can decide which method to use.

yalctay93 replied:
True! I'll make sure to add it.

functicons commented on docs/flink_job_cluster_guide.md:
````diff
@@ -0,0 +1,108 @@
+
+3. Use the following command to install the Flink job cluster chart:
+  ```bash
+  helm template --namespace example-namespace flink-job-cluster ./flink-job-cluster -f ./flink-job-cluster/values.yaml |kubectl apply -f -
````
Seems the `--namespace` flag doesn't work, the cluster is always deployed to the default namespace.

yalctay93 replied:
true, the --namespace flag should be used after the kubectl apply, basically like this:
```helm template  flink-job-cluster ./flink-job-cluster -f ./flink-job-cluster/values.yaml | kubectl apply --namespace foo -f - ```

yalctay93 replied:
Actually, I will change the documentation to deploy the cluter without helm template, since it has a disadvantage of not making use of the release feature that helm includes. Using the command ```helm --namespace=foo flink-job-cluster ./flink-job-cluster -f ./flink-job-cluster/values.yaml``` is the better option.

functicons commented on docs/flink_job_cluster_guide.md:
````diff
@@ -0,0 +1,108 @@
+
+3. Use the following command to install the Flink job cluster chart:
+  ```bash
+  helm template --namespace example-namespace flink-job-cluster ./flink-job-cluster -f ./flink-job-cluster/values.yaml |kubectl apply -f -
````
I got the following error about PodMonitor when running this command:

```
$ helm template --namespace foo flink-job-cluster ./flink-job-cluster -f ./flink-job-cluster/values.yaml | kubectl apply -f -

configmap/flink-job-cluster created
flinkcluster.flinkoperator.k8s.io/flink-job-cluster created
error: unable to recognize "STDIN": no matches for kind "PodMonitor" in version "monitoring.coreos.com/v1"
```

yalctay93 replied:
True, I will add into the prerequisites that a PodMonitor will require you to have a prometheus-operator running in the cluster along with its custom resource definitions. I believe you did not have it active in your cluster?

functicons commented on docs/flink_job_cluster_guide.md:
```diff
@@ -0,0 +1,108 @@
+
+## Check the Flink job cluster
+
+After using the helm template command, the following resources will be deployed
```
After running the command, I noticed only a JM is running, no TM and job.

```
$ kpods

NAMESPACE               NAME                                                        READY   STATUS    RESTARTS   AGE
default                 flink-job-cluster-jobmanager-7ccb67f784-ldr4s               1/1     Running   0          2m
flink-operator-system   flink-operator-controller-manager-85cffbc4fb-wvkwl          2/2     Running   0          20m
```

In the operator pod log, I noticed this error:

```
2020-08-03T16:50:15.777Z        ERROR   controllers.FlinkCluster        Failed to reconcile     {"cluster": "default/flink-job-cluster", "error": "Deployment.apps \"flink-job-cluster-taskmanager\" is invalid: spec.template.spec.containers[0].volumeMounts[0].name: Not found: \"config\""}
```

yalctay93 replied:
I am going to remove the configmap mount since it was needed for our specific purposes. I'll leave in the possibility to mount files. I believe it is up to the users that use the chart whether they want to mount something or not.

functicons commented on docs/flink_job_cluster_guide.md:
```diff
@@ -0,0 +1,108 @@
+
+## Run Jobs with InitContainer
+
+You have the option to download blob containers, which would be your JAR files to execute as jobs, directly into the Flink job cluster pods.
```
I don't know what you mean by "block containers", is it a term of Azure? I think we might want to avoid vendor-specific stuff here. Just remind the user that they can download job jars from the remote with init containers, but we don't need to go into vendor-specific technology.

yalctay93 replied:
Yes, it is a term of Azure, at least we did do it with azure blob containers. Sure, we can leave out vendor-specific technology.

yalctay93 commented:
> A few more comments, thanks!

I've just commited the changes. Can you retry the helm deployment? I have tried it on my local k8s cluster and it worked as expected. 

```flink-job-cluster-job-6xr5r                         0/1     Error     0          102m
flink-job-cluster-jobmanager-85cc48fdc9-d47s4       1/1     Running   0          103m
flink-job-cluster-taskmanager-7787cf998c-hlbv9      1/1     Running   0          103m
flink-job-cluster-taskmanager-7787cf998c-psrvv      1/1     Running   0          103m
flink-operator-controller-manager-9676cfbcf-86565   2/2     Running   0          4d
```
The error on the job says 

```
Could not build the program from JAR file.
```
that is because I do not have a JAR file provided in the Flink 1.9.3 image. It'd run if I had one, but for the test run I didnt want to use our project JAR files.

functicons commented on docs/flink_job_cluster_guide.md:
````diff
@@ -0,0 +1,133 @@
+
+### Why use the Helm Chart?
+- Using this Chart enables you to easily configure many Flink job clusters based on a single template
+- It enables you to use configuration manifests that belong to each other from a central place e.g. ```PodMonitor```
````
I don't quite understand what this means, could you rephrase?

yalctay93 replied:
I added a few more lines, to be honest in the current state of the Flink job cluster YAML, you could very well deploy it without the helm chart. However, as more manifest files are added that are deployed togheter with the Flink job cluster, such as the PodMonitor, it'll become easier to just enable them or configure them in one values.yaml. The biggest pro for us are the pre deployment checks that are included in helm as well as the post deployment check and automated rollback with --atomic.

functicons commented on docs/flink_job_cluster_guide.md:
````diff
@@ -0,0 +1,133 @@
+
+3. Use the following command to dry-run the Flink job cluster chart:
+  ```bash
+  helm install --dry-run --namespace=foo flink-job-cluster ./flink-job-cluster -f ./flink-job-cluster/values.yaml
````
s/foo/\<namespace\> here and below.

yalctay93 replied:
Done!

functicons commented on helm-chart/flink-job-cluster/values.yaml:
```diff
@@ -0,0 +1,113 @@
+job:
+  # job will look for a JAR file at /JARFiles/<JAR_FILE> and execute it
+  # className has to be valid and used in the provided JAR File
+  jarFile: /JARfiles/JARfile.jar
```
Can we replace this with `./examples/streaming/WordCount.jar` which exists in the container, so that the example could run successfully?

yalctay93 replied:
Sure! I've replaced it in the user guide and values.yaml, the result is now better:
```
NAME                                                READY   STATUS    RESTARTS   AGE
flink-job-cluster-job-mnxl4                         1/1     Running   0          17s
flink-job-cluster-jobmanager-85cc48fdc9-mfpvt       1/1     Running   0          59s
flink-job-cluster-taskmanager-7787cf998c-zstrj      1/1     Running   0          59s
flink-job-cluster-taskmanager-7787cf998c-zz4hh      1/1     Running   0          59s
flink-operator-controller-manager-9676cfbcf-86565   2/2     Running   0          4d21h
```

functicons commented on docs/flink_job_cluster_guide.md:
````diff
@@ -0,0 +1,133 @@
+- [Helm](https://helm.sh/docs/helm/helm_install/) version 2.x or 3.x
+
+Optional:
+- [Prometheus-Operator](https://github.com/helm/charts/tree/master/stable/prometheus-operator) to use the custom resource ```PodMonitor``` in order to scrape flink-job-cluster metrics
````
Can we add a link to this guide https://github.com/GoogleCloudPlatform/flink-on-k8s-operator/blob/master/docs/user_guide.md#monitoring-with-prometheus?

yalctay93 replied:
Done!

yalctay93 commented:
> I tested, the problems are solved, thanks! Left some minor comments.

Thank you! I've adjusted the user guide and values.yaml according to the comments you've made.

functicons commented:
/gcbrun
